### PR TITLE
Fixed ZCA n_components, and added tests

### PR DESCRIPTION
**Pull request by harmdevries89**

### CODE REVIEW

File: pylearn2/datasets/tests/test_preprocessing.py [hunk 1]
```diff
@@ -194,66 +194,124 @@ def test_rgb_yuv():
...
-    Calculates the inverse of X with numpy.linalg.inv if inv_P_ is not stored
-    """
-
-    def test(store_inverse):
-        rng = np.random.RandomState([1, 2, 3])
-        X = as_floatX(rng.randn(15, 10))
-        preprocessed_X = copy.copy(X)
-        preprocessor = ZCA(store_inverse=store_inverse)
+    def is_almost_equal(self, A, B, epsilon=1e-4):
+        abs_difference = np.abs(A - B)
```
Comment (dwf) [inline pylearn2/datasets/tests/test_preprocessing.py]: You can use `assert_allclose` with an `atol` argument for this, no need to reinvent it.

File: pylearn2/datasets/preprocessing.py [hunk 2]
```diff
@@ -1421,15 +1427,16 @@ def fit(self, X):
...
         if self.n_components:
-            eigs = eigs[:self.n_components]
-            eigv = eigv[:, :self.n_components]
+            eigs = eigs[-self.n_components:]
+            eigv = eigv[:, -self.n_components:]
 
-        if self.n_drop_components:
+        if self.n_drop_components and not self.n_components:
             eigs = eigs[self.n_drop_components:]
             eigv = eigv[:, self.n_drop_components:]
```
Comment (dwf) [inline pylearn2/datasets/preprocessing.py]: Might be a good idea to raise an error if both `n_components` and `n_drop_components` are not `None`.

File: pylearn2/datasets/tests/test_preprocessing.py [hunk 3]
```diff
@@ -194,66 +194,124 @@ def test_rgb_yuv():
...
+                                           preprocessor.inv_P_), identity)
+
+        preprocessor = ZCA(filter_bias=0.0)
+        preprocessed_X = self.get_preprocessed_data(preprocessor)
+
+        #Check if preprocessed data matrix is white
+        assert self.is_almost_equal(np.cov(preprocessed_X.transpose(),
+                                           bias=1), identity)
+
+        #Check if we obtain correct solution
```
Comment (dwf) [inline pylearn2/datasets/tests/test_preprocessing.py]: Lines 244 and 248 here are causing the build to fail (style guide says "#" has to have a space after it.)

**Comment by dwf** [comment]
Okay, this is ready pending the build fix and preferably using `numpy.testing.assert_allclose` instead of the homemade thing.

**Comment by lamblin** [comment]
The tolerance (at least the absolute one) of the tests should be relaxed when `floatX` is `float32`, as I have the following errors when running the tests in float32:
```
======================================================================
ERROR: Confirm that ZCA.inv_P_ is the correct inverse of ZCA.P_.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Tmp/lisa/os_v3/canopy/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/u/lamblinp/code/Pylearn2/pylearn2/datasets/tests/test_preprocessing.py", line 242, in test_zca
    bias=1), identity, rtol=1e-4)
  File "/u/lamblinp/code/Theano/theano/tests/unittest_tools.py", line 344, in assert_allclose
    raise WrongValue(val1, val2, rtol, atol)
WrongValue: WrongValue
           : shape, dtype, strides, min, max, n_inf, n_nan:
  Expected : (5, 5) float64 (40, 8) -2.92256885134e-05 0.999995412273 0 0
  Value    : (5, 5) float32 (20, 4) 0.0 1.0 0 0
  expected    : [[  9.99995412e-01   1.02717857e-05   1.30899588e-05  -2.03469480e-05
    7.78692182e-06]
 [  1.02717857e-05   9.99976268e-01  -2.92256885e-05   4.73881988e-05
   -1.73684807e-05]
 [  1.30899588e-05  -2.92256885e-05   9.99965396e-01   5.46143997e-05
   -2.18247959e-05]
 [ -2.03469480e-05   4.73881988e-05   5.46143997e-05   9.99910266e-01
    3.45030273e-05]
 [  7.78692182e-06  -1.73684807e-05  -2.18247959e-05   3.45030273e-05
    9.99987522e-01]]
  value    : [[ 1.  0.  0.  0.  0.]
 [ 0.  1.  0.  0.  0.]
 [ 0.  0.  1.  0.  0.]
 [ 0.  0.  0.  1.  0.]
 [ 0.  0.  0.  0.  1.]]
  Max Abs Diff:  8.97335485428e-05
  Mean Abs Diff:  2.71190267566e-05
  Median Abs Diff:  2.18247959077e-05
  Std Abs Diff:  1.9192050121e-05
  Max Rel Diff:  1.0
  Mean Rel Diff:  0.800003302805
  Median Rel Diff:  1.0
  Std Rel Diff:  0.399993394447

  rtol, atol: 0.0001 1e-05

-------------------- >> begin captured stdout << ---------------------
computing zca of a (8, 5) matrix
cov estimate took 0.00102591514587 seconds
eigh() took 7.10487365723e-05 seconds
computing zca of a (8, 5) matrix
cov estimate took 0.000457048416138 seconds
eigh() took 6.89029693604e-05 seconds

--------------------- >> end captured stdout << ----------------------

======================================================================
ERROR: Calculates the inverse of X with numpy.linalg.inv
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Tmp/lisa/os_v3/canopy/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/u/lamblinp/code/Pylearn2/pylearn2/datasets/tests/test_preprocessing.py", line 295, in test_zca_inverse
    test(store_inverse=True)
  File "/u/lamblinp/code/Pylearn2/pylearn2/datasets/tests/test_preprocessing.py", line 293, in test
    assert_allclose(self.X, preprocessor.inverse(preprocessed_X))
  File "/u/lamblinp/code/Theano/theano/tests/unittest_tools.py", line 344, in assert_allclose
    raise WrongValue(val1, val2, rtol, atol)
WrongValue: WrongValue
           : shape, dtype, strides, min, max, n_inf, n_nan:
  Expected : (8, 5) float64 (40, 8) -37.0 26.0 0 0
  Value    : (8, 5) float64 (40, 8) -36.9999694824 25.9999828339 0 0
  expected    : [[-10.   3.  19.   9. -15.]
 [  7.  26.  26.  26.  -3.]
 [ 17. -17. -37. -36. -11.]
 [ 19.  15.  -2.   5.   9.]
 [ -3.  -8. -35. -25.  -8.]
 [-18.   3.   4.  15.  14.]
 [  5.  -4.  -5.  -7. -11.]
 [ 23.  22.  15.  20.  12.]]
  value    : [[ -9.99999428   3.00000548  18.99999046   8.99999332 -14.9999876 ]
 [  7.          25.99998283  25.99998093  25.99998283  -3.00000095]
 [ 17.00000191 -16.9999752  -36.99996948 -35.9999733  -10.99999905]
 [ 18.99999523  14.99998665  -2.00000544   4.9999938    8.99998951]
 [ -3.00000286  -7.9999876  -34.99998093 -24.99998474  -7.99999952]
 [-18.00000763   2.99999571   3.9999938   15.00000191  14.00000381]
 [  5.0000056   -3.99998665  -4.99999237  -6.99999189 -10.99999237]
 [ 23.00000191  21.99998283  14.99998093  19.99998474  11.99999237]]
  Max Abs Diff:  3.0517578125e-05
  Mean Abs Diff:  9.9141150713e-06
  Median Abs Diff:  7.62939453125e-06
  Std Abs Diff:  7.50701022963e-06
  Max Rel Diff:  1.66893283904e-06
  Mean Rel Diff:  4.40295009714e-07
  Median Rel Diff:  3.70873482816e-07
  Std Rel Diff:  3.41323219874e-07

  rtol, atol: 1e-06 1e-08

-------------------- >> begin captured stdout << ---------------------
computing zca of a (8, 5) matrix
cov estimate took 0.000563859939575 seconds
eigh() took 6.60419464111e-05 seconds

--------------------- >> end captured stdout << ----------------------
```

**Comment by dwf** [comment]
Weird, why did Travis succeed?

**Comment by lamblin** [comment]
I assumed that it was using a dataset not present on Travis, but that's actually not the case.
For some reason, that test did not seem to be executed with Python 3 (which is the version executing the tests with float32).
The [Python 2 version](https://s3.amazonaws.com/archive.travis-ci.org/jobs/53822977/log.txt) says:
```
$ export PART="--exclude-dir=pylearn2/models"
Ran 551 tests in 850.200s
OK (SKIP=78)
```
[Python 3](https://s3.amazonaws.com/archive.travis-ci.org/jobs/53822980/log.txt):
```
$ export PART="--exclude-dir=pylearn2/models"
Ran 361 tests in 955.660s
OK (SKIP=38)
```

**Comment by lamblin** [comment]
Maybe because `class testZCA:` does not inherit from `object` and is not recognized as a test case?

**Comment by dwf** [comment]
Bah, I should have caught that.

But class testZCA: on Python 3 _does_ inherit from object, implicitly...
that's one of the changes, old-style classes are history, new-style is
king. So wtf gives?

**Comment by lamblin** [comment]
Or maybe now they _have_ to inherit from `TestCase`? Or the rule for picking up test cases by the name of the class is stricter?

**Comment by lamblin** [comment]
I opened https://github.com/lisa-lab/pylearn2/issues/1458

**Pull request merged**